hoomd/hpmc/UpdaterShape.h

```diff
@@ -0,0 +1,486 @@
+                                               access_mode::readwrite);
+
+            hoomd::RandomGenerator rng_i(hoomd::Seed(hoomd::RNGIdentifier::UpdaterShapeConstruct,
+                                                     timestep,
+                                                     this->m_sysdef->getSeed()),
+                                         hoomd::Counter(typ_i, 0, i_sweep));
+
+            // perform an in-place shape update on shape_param_new
+            m_move_function->update_shape(timestep,
+                                          m_step_size[typ_i],
+                                          typ_i,
+                                          shape_param_new,
+                                          rng_i);
+
+            // update det(I)
+            detail::MassProperties<Shape> mp(shape_param_new);
+            h_det.data[typ_i] = mp.getDetInertiaTensor();
+
+            m_exec_conf->msg->notice(5) << " UpdaterShape I=" << h_det.data[typ_i] << ", "
+                                        << h_det_old.data[typ_i] << std::endl;
+
+            // consistency check
+            assert(h_det.data[typ_i] != 0 && h_det_old.data[typ_i] != 0);
+
+            // compute log_boltz factor
+            log_boltz
+                = (*m_move_function)(timestep,             // current timestep
+                                     h_ntypes.data[typ_i], // number of particles of type typ_i,
+                                     typ_i,                // the type id
+                                     shape_param_new,      // new shape parameter
+                                     h_det.data[typ_i],    // new determinant_inertia_tensor
+                                     shape_param_old,      // old shape parameter
+                                     h_det_old.data[typ_i] // old determinant_inertia_tensor
+                );
+
+            // actually update the shape parameter in the integrator
+            m_mc->setParam(typ_i, shape_param_new);
+
+            // check if at least one overlap was caused
+            bool overlaps = static_cast<bool>(m_mc->countOverlaps(true));
+            // automatically reject if there are overlaps
+            if (overlaps)
+                {
+                m_exec_conf->msg->notice(5)
+                    << "UpdaterShape move rejected -- overlaps found" << std::endl;
+                // revert shape parameter changes
+                h_det.data[typ_i] = h_det_old.data[typ_i];
+                m_mc->setParam(typ_i, shape_param_old);
+                }
+            else // potentially accept, check Metropolis first
+                {
+                Scalar p = hoomd::detail::generate_canonical<Scalar>(rng_i);
+                Scalar Z = slow::exp(log_boltz);
+
+#ifdef ENABLE_MPI
+                if (m_multi_phase)
+                    {
+                    std::vector<Scalar> Zs;
+                    all_gather_v(Z, Zs, m_exec_conf->getMPICommunicator());
```

Comment:
```suggestion
                    all_gather_v(Z, Zs, m_exec_conf->getHOOMDWorldMPICommunicator());
```
Multi-phase needs to gather across the partitions, not the local domain.

Reply:
Hmm, that makes sense, I thought the code run two simulations in the same partition, but that doesn't make sense. However, in this case would this work as expected, because we would want to gather only on the root ranks of the partitions right?

Reply:
Yes, as written the multiplication of Z would produce the wrong results when using more than one rank in each partition.

Also, I think the all_gather for accept should be outside the `if (accept)` block. That call will deadlock as is when some ranks don't accept and others do.